### ISSUE TITLE
AMR performance 7/N: Simplying communication routines

### ISSUE THREAD

**Issue by pgrete** [parthenon-hpc-lab/parthenon#421]
There exist some call in the logic of our communication routine that I don't fully follow (and which may be leftover from some initial design ideas).
For example, the `resetBoundary()` function for each `Variable` resets the `var_cc` data pointer.
I'm not sure, this is still required and if the overall communication logic could not be simplified.

@jdolence I've been tagging you here as discussed last Wednesday as (IIRC) you're most familiar with these pieces/the original design.

**Comment by pgrete** [comment]
Looks like @Yurlungur may also have some good input on this issue.
Just found this comment in `variable.cpp` while trying to figure out what the difference between `bvars` and `bvars_main_int` is and if this is still need in Parthenon (or a remnant of the GRMHD code in Athena++)
```c++
  // TODO(JMM): This means RestrictBoundaries()
  // is called on EVERY stage, regardless of what
  // stage needs it.
  // The fix is to refactor BoundaryValues
  // to expose calls at either the `Variable`
  // or `MeshBlockData` and `MeshData` level.
  pmb->pbval->bvars.push_back(vbvar);
  pmb->pbval->bvars_main_int.push_back(vbvar)
```

**Comment by Yurlungur** [comment]
Yes. Let me refresh my memory on how this works. There's some optimization here that needs to be done.

**Comment by pgrete** [comment]
Three more riddles ;)
### 1 `coarse_s`
Why does `  ParArrayND<T> coarse_s; // used for sending coarse boundary calculation`, see variable.cpp
```c++
  if (pmb->pmy_mesh->multilevel)
    coarse_s = ParArrayND<T>(base_name + ".coarse", GetDim(6), GetDim(5), GetDim(4),
                             pmb->c_cellbounds.ncellsk(IndexDomain::entire),
                             pmb->c_cellbounds.ncellsj(IndexDomain::entire),
                             pmb->c_cellbounds.ncellsi(IndexDomain::entire));
```
have this shape?
This looks like it's covering the entire MeshBlock. Is this required? I'm not sure I fully follow.

### 2 `bvars_main_int`
Are ` bvars` and  `bvars_main_int` (see `bvals.hpp`)
```c++
  // variable-length arrays of references to BoundaryVariable instances
  // containing all BoundaryVariable instances:
  std::vector<std::shared_ptr<BoundaryVariable>> bvars;
  // subset of bvars that are exchanged in the main TimeIntegratorTaskList
  std::vector<std::shared_ptr<BoundaryVariable>> bvars_main_int
```
both still required in Parthenon or are they some remnant of some Athena++ feature?

### 3 RestrictBoundaries

In `bvals_refine.cpp` there's
```c++
// NOTE(JMM): ProlongateBounds has been split into RestrictBoundaries
// and ProlongateBoundaries,
// which are now called together as ProlongateBoundaries in
// `bvals/bondary_conditions.hpp`. This allows us to loop over all variables in a
// container.
void BoundaryValues::RestrictBoundaries() {
  std::shared_ptr<MeshBlock> pmb = GetBlockPointer();
  int &mylevel = pmb->loc.level;
  for (int n = 0; n < nneighbor; n++) {
    NeighborBlock &nb = neighbor[n];
    if (nb.snb.level >= mylevel) continue;

    // fill the required ghost-ghost zone
    IndexRange bni, bnj, bnk;
    ComputeRestrictionBounds_(nb, bni, bnj, bnk);

    // TODO(JMM): this loop should probably be a kokkos loop
    for (int nk = bnk.s; nk <= bnk.e; nk++) {
      for (int nj = bnj.s; nj <= bnj.e; nj++) {
        for (int ni = bni.s; ni <= bni.e; ni++) {
          int ntype = std::abs(ni) + std::abs(nj) + std::abs(nk);
          // skip myself or coarse levels; only the same level must be restricted
          if (ntype == 0 || nblevel[nk + 1][nj + 1][ni + 1] != mylevel) continue;

          // this neighbor block is on the same level
          // and needs to be restricted for prolongation
          RestrictGhostCellsOnSameLevel(nb, nk, nj, ni);
        }
      }
    }
  }
}
```
which has
```c++
if (nb.snb.level >= mylevel) continue;
```
and further down the following comment
```c++
// skip myself or coarse levels; only the same level must be restricted
```
If I read the first conditional correctly, the code that the comment refers to is never reached for `label = mylevel`, thus, restriction on the same level does never happen (while the comment says it should). Am I missing something?

**Comment by Yurlungur** [comment]
Ok well I can answer some of these without refreshing my memory. (I will get to the other stuff soon I promise. I was sick last week and I'm behind.)

> Why does `ParArrayND<T> coarse_s;` have this shape?

It is indeed covering the entire meshblock, at half resolution. It doesn't need to, of course. This wastes memory. But it's the implementation carried over from `Athena++`, and it's a single buffer, with the same topology of index space as the fine buffer, which is convenient. It's also not *much* wasted memory, since it's half the size of the fine buffer.

> Are `bvars` and `bvars_main_int` (see `bvals.hpp`) both still required in Parthenon or are they some remnant of some Athena++ feature?

I would be shocked if `bvars_main_int` did anything now. I think it's a holdover from Athena++'s tasking and we should remove it.

> If I read the first conditional correctly, the code that the comment refers to is never reached for `label = mylevel`, thus, restriction on the same level does never happen (while the comment says it should). Am I missing something?

Hmm... I don't understand what's going on here... I basically copy-pasted that code from an older routine but didn't examine it carefully. It's entirely possible that block of code doesn't do anything. We should check with a debugger or print statements or similar.
